docs/bounded_dynamism.md

```diff
@@ -0,0 +1,85 @@
+
+Bounded dynamism allows computations to encode the maximum runtime size of dynamic dimensions of tensors. This makes it possible to run particular computations on hardware that doesn’t support dynamic tensors and needs hints on the largest possible. For example, real time inference might require support for dynamic batch so that inference doesn’t need to wait for accumulation up to the static batch size. Similarly, during training the last batch in a dataset might be smaller than the max batch size if the number of examples in the dataset is not divisible by the batch size.
+
+
+# Proposal
+
+
+## (P1) Use RankedTensorType encoding field for bounds
+
+Bounds for a dynamic tensor are represented using the `TypeExtensionsAttr` using the encoding field in the `RankedTensorType`. Bounds in `TypeExtensionsAttr` is an `int64_t` array of size equal to rank of the tensor. Values corresponding to static dimensions must be `ShapedType::kDynamicSize` which is `-1`.
+
+For example, the following type represents a 2D tensor with up to `3` rows and exactly `5` columns.
+
+
+```
+tensor<?x5xf32, #stable_hlo.type_extensions<bounds = [3, -1]>>
+```
+
+
+Type compatibility in StableHLO also checks for compatibility of the bounds. Two types are compatible if there exists a runtime tensor that could match both the types. So, two types with different bounds are compatible but a type with bounds that is lower than the static size in the other type are not compatible. The above example type is compatible with `tensor<2x5xf32>` but not with `tensor<7x5xf32>` as it doesn’t respect the bound `3` on the first dimension.
+
+Currently, StableHLO dialect is using the MLIR core ranked tensor type to encode bounds. It should be noted that there is a plan to introduce a custom StableHLO type in the future that could natively support bounds along with custom pretty printing format. There will be a separate RFC describing all the details. Also, the proposal will follow StableHLO backward compatibility policies so it is safe to use this type currently.
+
+
+## (P2) StableHLO op semantics with bounded operands or results
+
+All ops that support dynamic operands or results can have bounds specified for them. However, the result types needs to be compatible with the inferrable result types using the operands and attributes. This allows result types to be more generic or specific as long as it is compatible with the inferred type.
+
+
+## (P3) get\_dimension\_size / set\_dimension\_size ops
```

Comment:
Under the hood, these two ops directly call their counterpart in the HLO layer. If not exactly, let's clarify what divergences we should be aware of.

Reply:
Yes, that is the case.